### PR TITLE
First implementation of Feature/edit reservation

### PR DESCRIPTION
**Pull request by JortWillemsen**

### CODE REVIEW

File: src/main/java/nl/asd/reservation/domain/Reservation.java [hunk 1]
```diff
@@ -82,10 +82,11 @@ public void reserveTimeslot(Timeslot timeslot, ReservationRepository repository)
         var reservations = repository.findByWorkplace(this.workplace);
         for (var reservation : reservations) {
             // Bepaal of de gewenste timeslot in de andere timeslots zit (conflicteert met)
-            var timeslots = reservation.slots;
-            if (timeslots.stream().anyMatch(t -> t.conflictsWith(timeslot))) {
+            if (timeslot.conflictsWith(reservation.slots)) {
                 throw new RuntimeException("This has already been reserved");
             }
+
+
```
Comment (xandervedder) [inline src/main/java/nl/asd/reservation/domain/Reservation.java:89]: Unnecessary newlines

File: src/test/java/nl/asd/reservation/application/ReservationServiceTest.java [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
     @BeforeEach
     public void initialize() {
-        this.service = new ReservationService(new FakeReservationRepository(), null);
+        this.repository = new FakeReservationRepository();
+        this.service = new ReservationService(this.repository, null);
+
+        var reservation = new Reservation(new ReservationId(1L), LocalDate.now(), ReservationType.ONCE, new WorkplaceId(1L));
+        reservation.reserveTimeslot(new Timeslot(LocalDateTime.now(), LocalDateTime.now().plusMinutes(30)), this.repository);
+        this.repository.save(reservation);
+
```
Comment (xandervedder) [inline src/test/java/nl/asd/reservation/application/ReservationServiceTest.java:27]: Remove newline

File: src/main/java/nl/asd/reservation/domain/Timeslot.java [hunk 3]
```diff
@@ -10,6 +11,16 @@ public long minutes() {
...
     }
 
+    // Checks if this slot conflicts with a number of different timeslots
+    public boolean conflictsWith(List<Timeslot> others) {
+        for(var slot : others) {
+            if(this.conflictsWith(slot)) {
+                return true;
+            }
+        }
+        return false;
```
Comment (xandervedder) [inline src/main/java/nl/asd/reservation/domain/Timeslot.java:21]:
```suggestion
        return others.map().anyMatch(Timeslot::conflictsWith);
```
Could maybe work

File: src/test/java/nl/asd/reservation/application/ReservationServiceTest.java [hunk 4]
```diff
@@ -26,4 +35,14 @@ public void shouldCreateReservationCorrectly() {
...
     }
+
+    @Test
+    public void shouldChangeWorkplaceCorrectly() {
+
+
+        var reservation = this.repository.ofId(new ReservationId(1));
+        this.service.changeWorkplace(reservation.getId(), new WorkplaceId(2));
+
+        assertEquals(new WorkplaceId(2), this.repository.ofId(reservation.getId()).getWorkplace());
```
Comment (xandervedder) [inline src/test/java/nl/asd/reservation/application/ReservationServiceTest.java:46]: Newline stuff here aswell and remember that this is only the happy path.

File: src/main/java/nl/asd/reservation/domain/Reservation.java [hunk 5]
```diff
@@ -96,4 +95,32 @@ public void reserveTimeslots(List<Timeslot> slots, ReservationRepository reposit
...
+        for(var slot : this.slots) {
+            if(slot.conflictsWith(other.slots)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // Change the workplace of this reservation.
+    public void changeWorkplace(WorkplaceId newWorkplaceId, ReservationRepository repository) {
```
Comment (xandervedder) [inline src/main/java/nl/asd/reservation/domain/Reservation.java:110]: Maybe we can rename this to `transferWorkplace` to adhere to the ubiquitous language? Change sounds very technical

File: src/main/java/nl/asd/reservation/application/ReservationService.java [hunk 6]
```diff
@@ -29,4 +29,11 @@ public ReservationId reserveWorkplace(WorkplaceId workplace, LocalDateTime from,
         // Mag dit?
         return reservation.getId();
     }
+
+    public ReservationId changeWorkplace(ReservationId id, WorkplaceId newWorkplaceId) {
+        var reservation = this.repository.ofId(id);
+        reservation.changeWorkplace(newWorkplaceId, repository);
```
[the remark this answers is not in the document]
Comment (Rubenvdbrink) [reply src/main/java/nl/asd/reservation/application/ReservationService.java:35]: We should do this with custom exception "WorkplaceNotFoundException"

[the remark this answers is not in the document]
Comment (Rubenvdbrink) [reply src/main/java/nl/asd/reservation/application/ReservationService.java:35]: Do this with custom exception "WorkplaceNotFoundException"

File: src/main/java/nl/asd/workplace/domain/Building.java [hunk 7]
```diff
@@ -53,6 +53,10 @@ public boolean includesWorkplace(WorkplaceId id) {
 
     public void registerWorkplace(Workplace workplace) {
         // TODO: business rules
```
Comment (xandervedder) [inline src/main/java/nl/asd/workplace/domain/Building.java:55]: Remove TODO?

File: src/main/java/nl/asd/workplace/domain/Workplace.java [hunk 8]
```diff
@@ -38,4 +39,19 @@ public int getFloor() {
...
+        return number == workplace.number && floor == workplace.floor && Objects.equals(id, workplace.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, number, floor);
+    }
 }
+
+
```
Comment (xandervedder) [inline src/main/java/nl/asd/workplace/domain/Workplace.java:57]: One newline too many?

File: src/main/java/nl/asd/reservation/application/ReservationService.java [hunk 9]
```diff
@@ -47,4 +47,16 @@ public void cancelReservation(ReservationId reservationId) {
 
         this.repository.delete(reservation);
     }
+
+    public ReservationId transferWorkplace(ReservationId id, WorkplaceId newWorkplaceId) {
```
[the remark this answers is not in the document]
Comment (xandervedder) [reply src/main/java/nl/asd/reservation/application/ReservationService.java]: MigrateTo seems better, yes

File: src/main/java/nl/asd/reservation/domain/Reservation.java [hunk 10]
```diff
@@ -116,6 +116,34 @@ public void reserveTimeslots(List<Timeslot> timeslots, ReservationRepository rep
...
 
+    // Checks if this reservation has any overlapping timeslots reserved on the to be compared to reservation.
+    public boolean conflictsWith(Reservation other) {
+        for (var slot : this.slots) {
+            if (slot.conflictsWith(other.slots)) {
+                return true;
+            }
+        }
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (xandervedder) [reply src/main/java/nl/asd/reservation/domain/Reservation.java:127]: Will check this later

File: src/main/java/nl/asd/reservation/application/ReservationService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (JortWillemsen) [reply src/main/java/nl/asd/reservation/application/ReservationService.java]: Would "migrate" be good ubiquitous language?

**Pull request merged**